cmd/outline-ss-server/main.go

```diff
@@ -99,79 +93,136 @@ func (s *SSServer) startPort(portNum int) error {
 		return conn, err
 	}
 	go service.StreamServe(accept, tcpHandler.Handle)
-	go packetHandler.Handle(port.packetConn)
-	return nil
+	return listener, nil
 }
 
-func (s *SSServer) removePort(portNum int) error {
-	port, ok := s.ports[portNum]
-	if !ok {
-		return fmt.Errorf("port %v doesn't exist", portNum)
-	}
-	tcpErr := port.tcpListener.Close()
-	udpErr := port.packetConn.Close()
-	delete(s.ports, portNum)
-	if tcpErr != nil {
+func (s *SSServer) startDirectUDP(addr *net.UDPAddr, cipherList service.CipherList) (*net.UDPConn, error) {
+	packetConn, err := net.ListenUDP("udp", addr)
+	if err != nil {
 		//lint:ignore ST1005 Shadowsocks is capitalized.
-		return fmt.Errorf("Shadowsocks TCP service on port %v failed to stop: %w", portNum, tcpErr)
+		return nil, fmt.Errorf("Shadowsocks UDP service failed to start on address %v: %w", addr.String(), err)
 	}
-	logger.Infof("Shadowsocks TCP service on port %v stopped", portNum)
-	if udpErr != nil {
+	logger.Infof("Shadowsocks UDP service listening on %v", packetConn.LocalAddr().String())
+	packetHandler := service.NewPacketHandler(s.natTimeout, cipherList, s.m)
+	go packetHandler.Handle(packetConn)
+	return packetConn, nil
+}
+
+func (s *SSServer) start(addr net.Addr, cipherList service.CipherList) (io.Closer, error) {
+	switch t := addr.(type) {
```

Comment:
We should switch on the listener type instead. And it should be "stream" or "packet", since that defines the handler.